In [15]:
p_input = """7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9"""

In [31]:
with open('2-1.txt') as f:
    p_input=f.readlines()

In [16]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [52]:
def safety(lst):
    # numpy version
    diffs = np.diff(lst)
    incr = np.all((diffs >= 1) & (diffs <= 3))
    decr = np.all((diffs <= -1) & (diffs >= -3))
    safe = incr or decr
    return safe

In [61]:
def safety(lst):
    # standard python version
    diffs = [lst[i+1] - lst[i] for i in range(len(lst)-1)]
    incr = all(1 <= d <= 3 for d in diffs)
    decr = all(-3 <= d <= -1 for d in diffs)
    return incr or decr

In [62]:
%%timeit
total = 0
for level in p_input:
    safe = safety(level)
    if safe:
        total += 1
    else:
        for index in range(len(lst)):
            new_lst = lst[:index] + lst[index + 1:]
            new_safe = safety(new_lst)
            if new_safe:
                total += 1
                break

60.6 ms ± 9.66 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


Numpy:

    118 ms ± 17.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Standard:

    8.74 ms ± 383 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)

Overhead of creating and managing array objects probably isn't worth it for this small of a dataset. Let's try and create a bigger one to see if numpy is more effective at scale.

In [51]:
# Numpy
p_input = np.random.randint(1, 10, size=(10000, 50))

In [60]:
# Standard Python
import random 
p_input = [[random.randint(1, 9) for _ in range(50)] for _ in range(10000)]

Numpy:

    456 ms ± 60.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Standard python:

    60.6 ms ± 9.66 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Suprising.

Possible explanations:

- Large number of dyanamic array operations
- Small arrays